### PR TITLE
Followup to #80730 and add tests

### PR DESCRIPTION
**Pull request by jennybuckley**
**What type of PR is this?**
/kind feature
/priority important-soon
/cc @liggitt
/sig api-machinery
/wg apply

**What this PR does / why we need it**:
Needed to address some comments from @liggitt on #80730 and add tests for the proto marshaller for metav1.Fields

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by jennybuckley** [comment]
This is not an api change

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go [hunk 1]
```diff
@@ -27,13 +27,20 @@ import (
...
 type ProtoFields struct {
 	// Map is the representation used in the alpha version of this API
-	Map map[string]Fields `json:"-" protobuf:"bytes,1,rep,name=map"`
+	Map map[string]ProtoFields `json:"-" protobuf:"bytes,1,rep,name=map"`
 
 	// Raw is the underlying serialization of this object.
 	Raw []byte `json:"-" protobuf:"bytes,2,opt,name=raw"`
 }
 
+// MarshalJSON implements json.Marshaler
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go]: this is a bit misleading/fragile... it only works on legacy objects where Map is set and Raw is not

File: staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go [hunk 2]
```diff
@@ -232,7 +232,7 @@ func CompatibilityTestFuzzer(scheme *runtime.Scheme, fuzzFuncs []interface{}) *f
 			field := metav1.ManagedFieldsEntry{}
 			c.Fuzz(&field)
 			if field.Fields != nil {
-				field.Fields.Raw = []byte("{}")
+				field.Fields.Raw = []byte(`{"18":{"19":null}}`)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go]: where do we ensure these bytes are interpretable by structured-merge-diff, or that structured-merge-diff only produces bytes in this format?

File: staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go [hunk 3]
```diff
@@ -231,8 +231,8 @@ func CompatibilityTestFuzzer(scheme *runtime.Scheme, fuzzFuncs []interface{}) *f
 		func(f *[]metav1.ManagedFieldsEntry, c fuzz.Continue) {
 			field := metav1.ManagedFieldsEntry{}
 			c.Fuzz(&field)
-			if field.Fields != nil {
-				field.Fields.Raw = []byte("{}")
+			field.Fields = &metav1.Fields{
+				Raw: []byte(`{"18":{"19":null}}`),
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go]: is there a way to fuzz or populate the corresponding structured-merge-diff struct and marshal it to produce this?

Comment (jennybuckley) [reply staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go]: Yes, there's this: https://github.com/kubernetes-sigs/structured-merge-diff/blob/0c1d754dd6485aa156ff3c8c599079adac6ec82f/fieldpath/set_test.go#L30

Comment (jennybuckley) [reply staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go]: It's not deterministic but we could make a deterministic version

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go (hunk 1 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go]: Hmm true. I can't think of a case where Map and Raw would both be set, but we should probably handle that somehow

Comment (jennybuckley) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go]: or give an error

File: staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go]: a small deterministic version sounds good

**Comment by jennybuckley** [comment]
Closing this, since https://github.com/kubernetes/kubernetes/pull/81754 removed the special proto marshalling anyway

**Pull request closed without merging**